File: dynamic_programming/longest-repeating-subsequence.java
import java.util.*;
import java.lang.*;
import java.io.*;

class GFG {
	public static void main (String[] args) {
		Scanner sc = new Scanner(System.in);

    int t = sc.nextInt();
    while (t > 0) {
      t--;
      // read the string length, and the string
      int n = sc.nextInt();
      sc.nextLine();
      String s = sc.nextLine();
      // print the length of the longest repeating subsequence in s
      System.out.println(LRS(s, n));
    }
	}

  public static int LRS(String s, int n) {
    int[][] memo = new int[n+1][n+1];
    for (int i = 0; i <= n; i++) {
      for (int j = 0; j <= n; j++) {
        memo[i][j] = 0;
      }
    }

    for (int i = 1; i <= n; i++) {
      for (int j = 1; j <= n; j++) {
        if (s.charAt(i-1) == s.charAt(j-1) && i != j) {
          memo[i][j] = 1 + memo[i-1][j-1];
        } else {
          memo[i][j] = Math.max(memo[i][j-1], memo[i-1][j]);
        }
      }
    }

    return memo[n][n];
  }
}

File: dynamic_programming/longest-repeating-subsequence.py
def LRS(S, n):
    memo = [[0 for _ in range(n+1)] for _ in range(n+1)]

    for i in range(1, n+1):
        for j in range(1, n+1):
            # if same char but not at same index
            if S[i-1] == S[j-1] and i != j:
                memo[i][j] = 1 + memo[i-1][j-1]
            else:
                memo[i][j] = max(memo[i-1][j], memo[i][j-1])

    return memo[n][n]

for _ in range(int(input())):
    # read a strings S of length n
    n = int(input())
    S = input()
    # find the length of the longest repeated subsequence in S
    print(LRS(S, n))
